Userland/Libraries/LibUnicode/URL.cpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2023, Simon Wanner <simon@skyrising.xyz>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/String.h>
+#include <AK/URL.h>
+#include <LibUnicode/IDNA.h>
+
+// This file implements the Unicode domain conversions that can't be implemented in AK.
+
+namespace AK {
+
+// https://url.spec.whatwg.org/#concept-domain-to-ascii
+static ErrorOr<String> domain_to_ascii(StringView domain, bool be_strict)
+{
+    // 1. Let result be the result of running Unicode ToASCII with domain_name set to domain, UseSTD3ASCIIRules set to beStrict, CheckHyphens set to false, CheckBidi set to true, CheckJoiners set to true, Transitional_Processing set to false, and VerifyDnsLength set to beStrict. [UTS46]
+    // 2. If result is a failure value, domain-to-ASCII validation error, return failure.
+    auto result = TRY(Unicode::IDNA::to_ascii(
+        Utf8View(domain),
+        Unicode::IDNA::CheckHyphens::No,
+        Unicode::IDNA::CheckBidi::Yes,
+        Unicode::IDNA::CheckJoiners::Yes,
+        be_strict ? Unicode::IDNA::UseStd3AsciiRules::Yes : Unicode::IDNA::UseStd3AsciiRules::No,
+        Unicode::IDNA::TransitionalProcessing::No,
+        be_strict ? Unicode::IDNA::VerifyDnsLength::Yes : Unicode::IDNA::VerifyDnsLength::No));
+
+    // 3. If result is the empty string, domain-to-ASCII validation error, return failure.
+    if (result.is_empty())
+        return Error::from_string_literal("Empty domain");
+
+    // 4. Return result.
+    return result;
+}
+
+// https://url.spec.whatwg.org/#concept-host-parser
+URL URL::create_with_unicode_url(String const& url_string)
```

Comment:
Can this be a free function? Seems a bit weird to implement a member function of a class in AK outside of AK. (Or does this call private functions?)

Reply:
Yes. I kind of followed what String::to_lowercase and friends are doing, but since this is static and doesn't touch anything private, it can be a free function.